rfd/0199-hardware-key-agent.md

```diff
@@ -0,0 +1,685 @@
+func (h *HardwarePrivateKey) Public() crypto.PublicKey {
+  return h.ref.PublicKey
+}
+
+// Implements [crypto.Signer].
+func (h *HardwarePrivateKey) Sign(rand io.Reader, digest []byte, opts crypto.SignerOpts) ([]byte, error) {
+  return h.service.Sign(h.ref, rand, digest, opts)
+}
+
+// Implements [crypto.HardwareSigner].
+func (h *HardwarePrivateKey) GetAttestationStatement() *AttestationStatement {
+  return h.ref.AttestationStatement
+}
+
+// Implements [crypto.HardwareSigner].
+func (h *HardwarePrivateKey) GetPrivateKeyPolicy() PrivateKeyPolicy {
+  return h.ref.PrivateKeyPolicy
+}
+```
+
+### UX
+
+#### PIN caching
+
+The PIN caching portion of this RFD does not present any additional UX concerns.
+It is purely a UX benefitting change, removing the need for back-to-back PIN
+prompts.
+
+#### Agent clients
+
+##### Teleport Connect
+
+Teleport Connect is an ideal runner for the key agent because it:
+
+1. Is a long-lived process
+1. Provides a UI for touch/PIN prompts, and can foreground itself for these prompts
+
+By default, if Teleport Connect detects a hardware key plugged in, it will automatically
+serve the hardware key agent service. This way, the agent will be served
+regardless of Teleport Connect's [login state](#running-the-agent-before-login).
+
+Note: the plugged in hardware key is detected by
+[listing smart cards available via the PC/SC interface](https://pkg.go.dev/github.com/go-piv/piv-go/piv#Cards).
+
+For example, if a user primarily wants to use `tsh`, but get PIN caching and
+PIN prompts in Teleport Connect, they could just launch Teleport Connect without
+logging in. Teleport Connect would just foreground itself with hardware key
+prompts for the user as needed without adding additional overhead.
+
+Alternative: By default, set `teleport.agent=false`. Once the user logs into
+Connect with a hardware key requirement for the first time, flip the flag to
+true indefinitely. The benefit with this approach is that we won't run an
+unused agent by default for users not using hardware key support. On the other
+hand, it requires the user to log in at least once for the feature to work as
+described in the example above.
+
+If desired, the agent can be disabled manually with a config option:
+
+| Property | Default | Description |
+|----------|---------|-------------|
+| `hardwareKeyAgent.enabled` | `false` | Starts the hardware key agent automatically |
+
+##### `tsh agent`
+
+`tsh agent` will be made available as a hidden command, primarily for
+development. If in the future we get requests to fully support this command,
+we may make it a public command and make any necessary improvements.
+
+For now, we will not put exorbitant effort into providing a good UX with this command
+(e.g. stealing focus for hardware key prompts, re-login on cert expiration),
+but it should be fully functional.
+
+##### Agent already running
+
+If there is already a hardware key agent running at `$TEMP/.Teleport-PIV/agent.sock`,
```

Comment:
I wonder if Teleport Connect should be more greedy about being the agent. It's the best UX by far.

I'm worried about poor UX when someone accidentally and unknowingly runs the key agent via tsh or tctl just by running some command at the start of their day, and then having a hard time finding their PIN prompts and figuring out what process they need to kill in order to get Teleport Connect to take over as the agent.

Reply:
How would the user accidentally or unknowingly run the key agent with `tsh`? It's going to a hidden command: `tsh piv agent`.